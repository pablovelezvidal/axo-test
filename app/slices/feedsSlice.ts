import { createSlice } from '@reduxjs/toolkit';
// eslint-disable-next-line import/no-cycle
import { AppThunk } from '../store';

const SEARCH_URI = 'https://api.twitter.com/1.1/statuses/user_timeline.json';
const BEARER_TOKEN = process.env.BEARER_TOKEN;

type State = {
  loading: boolean;
  hasErrors: boolean;
  feeds: Array<any>;
};

const initialState: State = {
  loading: false,
  hasErrors: false,
  feeds: [],
};

const feedsSlice = createSlice({
  name: 'feeds',
  initialState: initialState,
  reducers: {
    getFeeds: (state) => {
      state.loading = true;
    },
    getFeedsSuccess: (state, { payload }) => {
      state.feeds = payload;
      state.loading = false;
      state.hasErrors = false;
    },
    getFeedsFailure: (state) => {
      state.loading = false;
      state.hasErrors = true;
    },
  },
});

export const {
  getFeeds,
  getFeedsSuccess,
  getFeedsFailure,
} = feedsSlice.actions;

export const feedsSelector = (state: State): any => state.feeds;

// Asynchronous thunk action
export const fetchFeeds = (
  screenName: string,
  quantity: number = 10
): AppThunk => {
  return async (dispatch: (arg0: any) => void) => {
    dispatch(getFeeds());

    try {
      const response = await fetch(
        `${SEARCH_URI}?screen_name=${screenName}&count=${quantity}`,
        {
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            Authorization: `Bearer ${BEARER_TOKEN}`,
          },
        }
      );
      const data = await response.json();
      dispatch(getFeedsSuccess(data));
    } catch (error) {
      dispatch(getFeedsFailure());
    }
  };
};

export default feedsSlice.reducer;
