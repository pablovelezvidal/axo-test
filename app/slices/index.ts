import feedsReducer from './feedsSlice';
import usersReducer from './usersSlice';

const reducers = {
  feeds: feedsReducer,
  users: usersReducer,
};

export default reducers;
