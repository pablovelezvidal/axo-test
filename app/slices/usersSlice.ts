import { createSlice } from '@reduxjs/toolkit';
// eslint-disable-next-line import/no-cycle
import { AppThunk } from '../store';

const SEARCH_URI =
  'https://typeahead-js-twitter-api-proxy.herokuapp.com/demo/search';

type State = {
  isLoading: boolean;
  hasErrors: boolean;
  users: Array<any>;
  userSelected: string;
};

const initialState: State = {
  isLoading: false,
  hasErrors: false,
  users: [],
  userSelected: '',
};

const usersSlice = createSlice({
  name: 'users',
  initialState,
  reducers: {
    getUsers: (state) => {
      state.isLoading = true;
    },
    getUsersSuccess: (state, { payload }) => {
      state.users = payload;
      state.isLoading = false;
      state.hasErrors = false;
    },
    getUsersFailure: (state) => {
      state.isLoading = false;
      state.hasErrors = true;
    },
    setUserSelected: (state, { payload }) => {
      state.userSelected = payload;
    },
  },
});

export const {
  getUsers,
  getUsersSuccess,
  getUsersFailure,
  setUserSelected,
} = usersSlice.actions;

export const usersSelector = (state: State): any => state.users;

// Asynchronous thunk action
export const fetchUsers = (query: string): AppThunk => {
  return async (dispatch: (arg0: any) => void) => {
    dispatch(getUsers());

    try {
      const response = await fetch(`${SEARCH_URI}?q=${query}`, {
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      });
      const items = await response.json();
      const users = items.map((i: any) => {
        return {
          profile_image_url_https: i.profile_image_url_https,
          id: i.id,
          screen_name: i.screen_name,
        };
      });
      dispatch(getUsersSuccess(users));
    } catch (error) {
      dispatch(getUsersFailure());
    }
  };
};

export default usersSlice.reducer;
