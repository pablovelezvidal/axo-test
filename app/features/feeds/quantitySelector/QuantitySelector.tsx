import React from 'react';

import { useSelector, useDispatch } from 'react-redux';

import { fetchFeeds } from '../../../slices/feedsSlice';
import { usersSelector } from '../../../slices/usersSlice';

export interface QuantitySelectorProps {}

const QuantitySelector: React.SFC<QuantitySelectorProps> = () => {
  const dispatch = useDispatch();
  const { userSelected } = useSelector(usersSelector);

  const handleChange = (event: any) => {
    dispatch(fetchFeeds(userSelected, event.target.value));
  };

  return (
    <select
      onChange={handleChange}
      className="browser-default custom-select"
      defaultValue="10"
    >
      <option value="10">Load 10 Tweets</option>
      <option value="20">Load 20 Tweets</option>
      <option value="50">Load 50 Tweets</option>
    </select>
  );
};

export default QuantitySelector;
