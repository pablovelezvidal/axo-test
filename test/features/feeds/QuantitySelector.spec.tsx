/* eslint react/jsx-props-no-spreading: off, @typescript-eslint/ban-ts-comment: off */
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import QuantitySelector from '../../../app/features/feeds/quantitySelector/QuantitySelector';
import * as usersSlice from '../../../app/slices/usersSlice';

Enzyme.configure({ adapter: new Adapter() });

function setup(
  preloadedState: any = {
    users: {
      isLoading: false,
      hasErrors: false,
      users: [],
      userSelected: 'User Selected',
    },
  }
) {
  const store = configureStore({
    reducer: { users: usersSlice.default },
    preloadedState,
  });

  const getWrapper = () =>
    mount(
      <Provider store={store}>
        <Router>
          <QuantitySelector />
        </Router>
      </Provider>
    );
  const component = getWrapper();
  return {
    store,
    component,
    select: component.find('select'),
  };
}

describe('quantity selector component should have default value of 10', () => {
  const { select } = setup();

  it('should display the title according to the user selected', () => {
    expect(select.props().defaultValue).toMatch(/^10$/);
  });

  it('should match exact snapshot', () => {
    const { store } = setup();
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <QuantitySelector />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
