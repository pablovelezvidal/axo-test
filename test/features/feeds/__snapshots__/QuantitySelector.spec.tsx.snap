// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quantity selector component should have default value of 10 should match exact snapshot 1`] = `
<select
  className="browser-default custom-select"
  defaultValue="10"
  onChange={[Function]}
>
  <option
    value="10"
  >
    Load 10 Tweets
  </option>
  <option
    value="20"
  >
    Load 20 Tweets
  </option>
  <option
    value="50"
  >
    Load 50 Tweets
  </option>
</select>
`;
